src/main/haskell/kore/test/Test/Kore/Step/Substitution.hs

```diff
@@ -0,0 +1,152 @@
+                 ( mkVar )
+import           Kore.Predicate.Predicate
+                 ( makeTruePredicate )
+import           Kore.Step.ExpandedPattern
+                 ( PredicateSubstitution (..) )
+import           Kore.Step.Substitution
+                 ( mergeAndNormalizeSubstitutions )
+import           Kore.Unification.Error
+import           Kore.Unification.Unifier
+                 ( UnificationSubstitution )
+import           Kore.Variables.Fresh.IntCounter
+                 ( runIntCounter )
+
+import qualified Test.Kore.IndexedModule.MockMetadataTools as Mock
+                 ( makeMetadataTools, makeSortTools )
+import qualified Test.Kore.Step.MockSymbols as Mock
+
+test_mergeAndNormalizeSubstitutions :: [TestTree]
+test_mergeAndNormalizeSubstitutions = give mockSortTools
+    [ testCase "Constructor normalization"
+        -- [x=constructor(a)] + [x=constructor(a)]  === [x=constructor(a, a)]
+        -- TODO(Vladimir) the result is x=constructor(a) instead of x=(constructor(a,a))
+        (assertEqual ""
+            ( Right
+                ( PredicateSubstitution
+                    makeTruePredicate
+                    [   ( Mock.x
+                        , Mock.constr10 Mock.a
+                        )
+                    ]
+                )
+            )
+            ( normalize
+                [   ( Mock.x
+                    , Mock.constr10 Mock.a
+                    )
+                ]
+                [   ( Mock.x
+                    , Mock.constr10 Mock.a
+                    )
+                ]
+            )
+        )
+    , testCase "Double constructor is bottom"
+        -- [x=constructor(a)] + [x=constructor(constructor(a))]  === bottom?
+        -- TODO(Vladimir) the result is error instead of bottom
+        (assertEqual ""
+            ( Left
+                ( UnificationError
+                    ( PatternClash
+                        ( HeadClash Mock.aSymbol )
+                        ( HeadClash Mock.constr10Symbol )
+                    )
+                )
+            )
+            ( normalize
+                [   ( Mock.x
+                    , Mock.constr10 Mock.a
+                    )
+                ]
+                [   ( Mock.x
+                    , Mock.constr10 (Mock.constr10 Mock.a)
+                    )
+                ]
+            )
+        )
+    , testCase "Constructor and constructor of function errors"
+        -- [x=constructor(a)] + [x=constructor(f(a))]  === error
+        (assertEqual ""
+            ( Left (UnificationError (NonConstructorHead Mock.fSymbol)) )
+            ( normalize
+                [   ( Mock.x
+                    , Mock.constr10 Mock.a
+                    )
+                ]
+                [   ( Mock.x
+                    , Mock.constr10 (Mock.f Mock.a)
+                    )
+                ]
+            )
+        )
+    , testCase "Constructor circular dependency?"
+        -- [a=b] + [b=constructor(a)]  === bottom
```

Comment:
Huh... There are a few things broken here (not by you).

First, could you delete `substitutionToStepError`? It is not used anywhere.

Second, could you call `ctorSubstitutionCycleToBottom` in `normalizeSubstitutionAfterMerge`? It should fix this test. I guess then you could delete its usage from `BaseStep.hs`.

Actually, even better, I guess you could fully delete the `CtorCircularVariableDependency` error case and you could make `normalizeSubstitution` return bottom in this case.

Reply:
How would you feel about renaming `CtorCircularVariableDependency vars` to a unit ctor `Bottom`, and mapping it to `Right bottom` in `normalizeSubstitutionAfterMerge`?

Reply:
It would work, I guess, but it would be less clean (i.e. it would be obvious that you're handling a non-error case as an error, and you couldn't even pretend that it would be useful somewhere as an actual error) why would that be better than handling this in `normalizeSubstitution`